src/core/algorithms/nar/value_range.cpp

```diff
@@ -0,0 +1,109 @@
+#include "value_range.h"
+
+namespace model {
+
+StringValueRange::StringValueRange(model::TypedColumnData const& column) {
+    domain = std::vector<std::string>();
+    for (size_t row_index = 0; row_index < column.GetNumRows(); row_index++) {
+        std::byte const* value = column.GetValue(row_index);
+        std::string string_value = model::Type::GetValue<std::string>(value);
+        bool first_occurrence =
+                std::find(domain.begin(), domain.end(), string_value) == domain.end();
+        if (first_occurrence) {
+            domain.emplace_back(string_value);
+        }
+    }
+}
+
+std::string StringValueRange::ToString() const {
+    std::string result;
+    result += "[";
+    if (domain.size() > 0) {
+        result += domain[0];
+    }
+    for (size_t i = 1; i < domain.size(); i++) {
+        result += (", " + domain[i]);
+    }
+    result += "]";
+    return result;
+}
+
+std::string IntValueRange::ToString() const {
+    std::string result;
+    result += "[";
+    result += std::to_string(lower_bound);
+    result += " - ";
+    result += std::to_string(upper_bound);
+    result += "]";
+    return result;
+}
+
+IntValueRange::IntValueRange(model::TypedColumnData const& column) {
+    bool initialized = false;
+    for (size_t row_index = 0; row_index < column.GetNumRows(); row_index++) {
+        std::byte const* value = column.GetValue(row_index);
+        long int int_value = model::Type::GetValue<model::Int>(value);
+        if (!initialized) {
+            lower_bound = int_value;
+            upper_bound = int_value;
+            initialized = true;
+            continue;
+        }
+        if (int_value > upper_bound) {
+            upper_bound = int_value;
+        }
+        if (int_value < lower_bound) {
+            lower_bound = int_value;
+        }
+    }
+}
+
+DoubleValueRange::DoubleValueRange(model::TypedColumnData const& column) {
+    bool initialized = false;
+    for (size_t row_index = 0; row_index < column.GetNumRows(); row_index++) {
+        std::byte const* value = column.GetValue(row_index);
+        double double_value = model::Type::GetValue<model::Double>(
+                value);  // is it okay to use a non-long double here? clion says GetValue returns
+                         // just double.
+        if (!initialized) {
+            lower_bound = double_value;
+            upper_bound = double_value;
+            initialized = true;
+            continue;
+        }
+        if (double_value > upper_bound) {
+            upper_bound = double_value;
+        }
+        if (double_value < lower_bound) {
+            lower_bound = double_value;
+        }
+    }
+}
```

Comment:
This method repeats `IntValueRange::IntValueRange`. The only difference is `int_value`/`double_value` type. I think, you can use template function like
```
template <typename T>
CreateValueRange(...) {
    ...
    T typed_value = model::Type::GetValue<T>(value);
    ...
```
and in constructor will be `CreateValueRange<model::Int>(...)` and `CreateValueRange<model::Double>(...)`.